lib/chef/client.rb

```diff
@@ -637,16 +652,19 @@ def node_name
     # @api private
     #
     def register(client_name = node_name, config = Chef::Config)
-      if Chef::HTTP::Authenticator.detect_certificate_key(client_name)
-        if File.exists?(config[:client_key])
-          logger.warn("WARNING - Client key #{client_name} is present on disk, ignoring that in favor of key stored in CertStore")
-        end
-        events.skipping_registration(client_name, config)
-        logger.trace("Client key #{client_name} is present in certificate repository - skipping registration")
-        config[:client_key] = "Cert:\\LocalMachine\\My\\chef-#{client_name}"
-      elsif !config[:client_key]
+      if !config[:client_key]
         events.skipping_registration(client_name, config)
         logger.trace("Client key is unspecified - skipping registration")
+      elsif ::Chef::Config[:migrate_key_to_keystore] == true && ChefUtils.windows?
+        cert_name = "chef-#{client_name}"
+        result = check_certstore_for_key(cert_name)
+        if result.rassoc("#{cert_name}")
+          logger.trace("Client key #{config[:client_key]} is present in Certificate Store - skipping registration")
+        else
+          move_key_and_register(cert_name)
+          logger.trace("Client key #{config[:client_key]} moved to the Certificate Store - skipping registration")
```

Comment:
This is some voodoo we have to do -because of the nature of how public/private keys get stored, we need to have a PFX blob to do that. The only want to do that currently is OpenSSL / PowerShell. By going this route, we essentially abandon the client.pem or validator.pem on disk and create a new public key for the node.